### ISSUE TITLE
Dummy data not generating expected incidences

### ISSUE THREAD

**Issue by alexwalkercebm** [ebmdatalab/opensafely-research-template#94]
The csv [here](https://gist.github.com/alexwalkercebm/8de0c7ec7c89c3184aae5f8d9ddff5bb) is generated from the ICS repo at this point https://github.com/ebmdatalab/opensafely-ics-research/tree/feac87418988a81cbedf92b8c14cb22ec4fc54d4 (though I expect it would be similar in the most recent version). I've removed a lot of the columns manually for clarity.

All of the incidences should be set to 0.2 by the default, however:

- The outcome columns, which are restricted to `on_or_after="2020-03-01"` only have values at the very top of the csv, with e.g. only 3/1000 outcomes in the `icu_date_admitted` column.

- The drug covariates, which are restricted by `between=["2019-11-01", "2020-03-01"]` have values slightly lower down the csv, but still only appear in e.g. 6/1000 for `high_dose_ics`.

- Smoking status, which is restricted by `on_or_before="2020-03-01"` has many more values (197/1000)

It seems to me that the underlying population has an incidence of 0.2, but this is then restricted to much lower by the date restrictions.

**Comment by alexwalkercebm** [comment]
This is not causing an issue for the descriptive stuff, but will likely cause issues soon for modelling, as it will mean that only a handful of drug exposed people have any outcome events.

**Comment by sebbacon** [comment]
So, there is a sort-of bug to fix here, but there's also an open question about what the right behaviour is.

As you surmise I'd implemented "20% incidence and a date filter of after 1st March" as "the incidence in the population is 20%" -- but "population" in this case refers to the expectations defined, not the properties of the underlying population. So for example if a covariate has a date expectation of `{"earliest": "1900-01-01", "latest": "2020-04-01"}` then the incidence is applied across that, and then the date filter on top.

In this model the correct fix, if you wanted 20% of the population to be flagged within the `between` dates, would be to set the expectation dates to match the between dates.

However I can see this is a bit awkward. Sleeping on it might give me some insights into a better syntax for this -- suggestions welcome.

**Comment by alexwalkercebm** [comment]
The nicely distributed exponentially increasing dates are a neat feature, but for our current purposes may not be necessary?

All we really need for date columns is for them to randomly appear at the specified incidence, and be randomly distributed within the dates specified by `on_or_before` etc.

**Comment by alexwalkercebm** [comment]
Having looked at the code more, I think your suggestion to set the expectation dates to match the `between` dates is fine, but a uniform distribution would work better for that. I've made a PR for this here: https://github.com/ebmdatalab/opensafely-research-template/pull/99

If it would be possible to get the `earliest_date` and `latest_date` from the `between` dates, that would make things more convenient, but I'm not sure how to do that without diving much deeper.

**Comment by sebbacon** [comment]
I'm thinking the idea of automatically setting from `between` might be overkill - there aren't always betweens, (e.g. `on_or_after`) and I quite like being explicit in the expectations.  But happy to  be convinced otherwise?

**Comment by alexwalkercebm** [comment]
Yeah not critical, and probably not worth investing much time on now, unless we come to find adding in separate expectation date ranges especially onerous.
